demos/common/utilities/jws_using_mbedtls.c

```diff
@@ -0,0 +1,634 @@
+                                    uint32_t ulELength,
+                                    unsigned char * pucBuffer,
+                                    uint32_t ulBufferLength )
+{
+    AzureIoTResult_t xResult;
+    int32_t lMbedTLSResult;
+
+    if( ulBufferLength < jwsRSA3072_SIZE + jwsSHA256_SIZE )
+    {
+        LogInfo( ( "Buffer Not Large Enough\n" ) );
+        return 1;
+    }
+
+    unsigned char * pucShaBuffer = pucBuffer + jwsRSA3072_SIZE;
+    size_t ulDecryptedLength;
+
+    /* The signature is encrypted using the input key. We need to decrypt the */
+    /* signature which gives us the SHA256 inside a PKCS7 structure. We then compare */
+    /* that to the SHA256 of the input. */
+    mbedtls_rsa_context ctx;
+
+    mbedtls_rsa_init( &ctx, MBEDTLS_RSA_PKCS_V15, 0 );
+
+    LogInfo( ( "--- Initializing Decryption ---\n" ) );
+
+    lMbedTLSResult = mbedtls_rsa_import_raw( &ctx,
+                                             pucN, ulNLength,
+                                             NULL, 0,
+                                             NULL, 0,
+                                             NULL, 0,
+                                             pucE, ulELength );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_import_raw res: %i\n", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    lMbedTLSResult = mbedtls_rsa_complete( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_complete res: %i\n", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    lMbedTLSResult = mbedtls_rsa_check_pubkey( &ctx );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_check_pubkey res: %i\n", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    LogInfo( ( "--- Decrypting ---\n" ) );
+
+    /* RSA */
+    lMbedTLSResult = mbedtls_rsa_pkcs1_decrypt( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, &ulDecryptedLength, pucSignature, pucBuffer, jwsRSA3072_SIZE );
+
+    if( lMbedTLSResult != 0 )
+    {
+        LogError( ( "mbedtls_rsa_pkcs1_decrypt res: %i\n", lMbedTLSResult ) );
+        mbedtls_rsa_free( &ctx );
+        return lMbedTLSResult;
+    }
+
+    mbedtls_rsa_free( &ctx );
+
+    LogInfo( ( "---- Calculating SHA256 over input ----\n" ) );
+    xResult = prvJWS_SHA256Calculate( pucInput, ulInputLength,
+                                      pucShaBuffer );
+
+    LogInfo( ( "--- Checking if SHA256 of header+payload matches decrypted SHA256 ---\n" ) );
+
+    /* TODO: remove this once we have a valid PKCS7 parser. */
+    int doTheyMatch = memcmp( pucBuffer + jwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, jwsSHA256_SIZE );
+
+    if( doTheyMatch == 0 )
+    {
+        LogInfo( ( "SHA of JWK matches\n" ) );
+        xResult = 0;
+    }
+    else
+    {
+        LogError( ( "SHA of JWK does NOT match\n" ) );
```

Comment:
I mightttt agree but the logging here for that takes a bit of shuffling things around. Either some for loop printing it hex by hex or base64 encoding it. I guess if people really need to dive in, they are free to edit this file (since it's not API source code)